src/core/vector/qgsvectorlayerprofilegenerator.cpp

```diff
@@ -802,423 +814,451 @@ bool QgsVectorLayerProfileGenerator::generateProfileForPoints()
+      *extZOut++ = ( std::isnan( interpolatedPoint.z() ) ? 0.0 : interpolatedPoint.z() ) + extrusion;
 
-    if ( !mProfileCurveEngine->intersects( feature.geometry().constGet() ) )
-      continue;
+    mResults->mRawPoints.append( interpolatedPoint );
+    mResults->minZ = std::min( mResults->minZ, interpolatedPoint.z() );
+    mResults->maxZ = std::max( mResults->maxZ, interpolatedPoint.z() );
+    if ( mExtrusionEnabled )
+    {
+      mResults->minZ = std::min( mResults->minZ, interpolatedPoint.z() + extrusion );
+      mResults->maxZ = std::max( mResults->maxZ, interpolatedPoint.z() + extrusion );
+    }
 
-    mExpressionContext.setFeature( feature );
+    const double distance = mProfileCurveEngine->lineLocatePoint( interpolatedPoint, &lastError );
+    *outDistance++ = distance;
 
-    const QgsGeometry g = feature.geometry();
-    if ( g.isMultipart() )
-    {
-      for ( auto it = g.const_parts_begin(); it != g.const_parts_end(); ++it )
-      {
-        if ( !mProfileCurveEngine->intersects( *it ) )
-          continue;
+    mResults->mDistanceToHeightMap.insert( distance, interpolatedPoint.z() );
+    lastDistanceAlongProfileCurve = distance;
+  }
 
-        processCurve( feature, qgsgeometry_cast< const QgsCurve * >( *it ) );
-      }
+  // insert nan point to end the line
+  mResults->mDistanceToHeightMap.insert( lastDistanceAlongProfileCurve + 0.000001, qQNaN() );
+
+  if ( mFeedback->isCanceled() )
+    return;
+
+  if ( mExtrusionEnabled )
+  {
+    std::unique_ptr< QgsLineString > ring = std::make_unique< QgsLineString >( newX, newY, newZ );
+    std::unique_ptr< QgsLineString > extrudedRing = std::make_unique< QgsLineString >( newX, newY, extrudedZ );
+    std::unique_ptr< QgsLineString > reversedExtrusion( extrudedRing->reversed() );
+    ring->append( reversedExtrusion.get() );
+    ring->close();
+    transformedParts.append( QgsGeometry( new QgsPolygon( ring.release() ) ) );
+
+    std::unique_ptr< QgsLineString > distanceVHeightRing = std::make_unique< QgsLineString >( newDistance, newZ );
+    std::unique_ptr< QgsLineString > extrudedDistanceVHeightRing = std::make_unique< QgsLineString >( newDistance, extrudedZ );
+    std::unique_ptr< QgsLineString > reversedDistanceVHeightExtrusion( extrudedDistanceVHeightRing->reversed() );
+    distanceVHeightRing->append( reversedDistanceVHeightExtrusion.get() );
+    distanceVHeightRing->close();
+    crossSectionParts.append( QgsGeometry( new QgsPolygon( distanceVHeightRing.release() ) ) );
+  }
+  else
+  {
+    transformedParts.append( QgsGeometry( new QgsLineString( newX, newY, newZ ) ) );
+    crossSectionParts.append( QgsGeometry( new QgsLineString( newDistance, newZ ) ) );
+  }
+};
+
+void QgsVectorLayerProfileGenerator::processTriangleIntersectForPolygon( const QgsPolygon *sourcePolygon, const QgsPolygon *intersectionPolygon, QVector< QgsGeometry > &transformedParts, QVector< QgsGeometry > &crossSectionParts )
+{
+  bool oldExtrusion = mExtrusionEnabled;
```

Comment:
I don't understand this logic -- can you add comments describing why this is so?

Reply:
indeed...